### PR TITLE
Add WeChat API

### PR DESCRIPTION
**Pull request by xhusky**
Support for Wechat Open Platform

see [https://open.weixin.qq.com/cgi-bin/showdocument?action=dir_list&t=resource/res_list&verify=1&id=open1419316505&token=&lang=zh_CN](url)

### CODE REVIEW

File: scribejava-apis/src/main/java/com/github/scribejava/apis/WeChatApi20.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        if (scope != null) {
+            parameters.add(WeChatConstants.SCOPE, scope);
+        }
+
+        final String state = config.getState();
+        if (state != null) {
+            parameters.add(WeChatConstants.STATE, state);
+        }
+
+        return parameters.appendTo(getAuthorizationBaseUrl());
```
Comment (kullfar) [inline scribejava-apis/src/main/java/com/github/scribejava/apis/WeChatApi20.java]: pity. for single param name...
Maybe let it be like this:
```
final ParameterList parameters = new ParameterList(additionalParams);
parameters.add(WeChatConstants.CLIENT_ID, config.getApiKey());
return parameters.appendTo(super.getAuthorizationUrl(config, additionalParams));
```
or even
```
if (additionalParams==null) {
  additionalParams = Collections.singletonMap(WeChatConstants.CLIENT_ID, config.getApiKey());
} else {
  additionalParams.put(WeChatConstants.CLIENT_ID, config.getApiKey());
}
return super.getAuthorizationUrl(config, additionalParams);
```

File: scribejava-apis/src/main/java/com/github/scribejava/apis/service/WeChatOAuth20ServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        final DefaultApi20 api = getApi();
+        final OAuthRequest request = new OAuthRequest(api.getAccessTokenVerb(), api.getAccessTokenEndpoint());
+        final OAuthConfig config = getConfig();
+
+        request.addParameter(WeChatConstants.CLIENT_ID, config.getApiKey());
+        request.addParameter(WeChatConstants.CLIENT_SECRET, config.getApiSecret());
+        request.addParameter(WeChatConstants.CODE, code);
+        request.addParameter(WeChatConstants.GRANT_TYPE, WeChatConstants.AUTHORIZATION_CODE);
+
+        return request;
```
Comment (kullfar) [inline scribejava-apis/src/main/java/com/github/scribejava/apis/service/WeChatOAuth20ServiceImpl.java]: the same here.
final OAuthRequest request = super.createAccessTokenRequest(code);
request.addParameter...
return request;

File: scribejava-apis/src/main/java/com/github/scribejava/apis/wechat/WeChatToken.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    // Unique identifier for Authorized user
+    private final String openId;
+
+    // User's unified identifier for a WeChat open platform account
+    private final String unionId;
+
+    public WeChatToken(String accessToken, String openId, String unionId, String rawResponse) {
+        this(accessToken, null, null, null, null, openId, unionId, rawResponse);
+    }
```
Comment (kullfar) [inline scribejava-apis/src/main/java/com/github/scribejava/apis/wechat/WeChatToken.java]: Seems we don't need this constructor. Constructor in GoogleToken as well.

**Comment by kullfar** [comment]
Hi,
your pull request looks good. The only things I can't like are createAccessTokenRequest and getAuthorizationUrl methods overriden. But seems there is no better way to do this for the moment. So it's up to you. We can leave them as it is now. Not used Constructor in Token as well. We can leave it.
Add the line to the README.md with new API added please.

And I need test user and app credentials to run the example before merging to the master.

ps.Thanks!

**Comment by xhusky** [comment]
@kullfar 
Test user and app credentials has send to your email (kullfar@gmail.com).
Thanks.

**Comment by kullfar** [comment]
Provided credentials aren't working (maybe they are outdated already?)
I'm waiting for the new working test credentials to run the example and merge it to the master.

**Comment by kullfar** [comment]
There are problems with test user credentials. I have tried some times to register. No success. WeChat didn't send me the sms.
I still need test user credentials to run the example to merge pull request to the master.

**Comment by xhusky** [comment]
Another test user account has been sent to you email, please try again.

**Pull request closed without merging**